### ISSUE TITLE
Rails 4.2 Encoding::UndefinedConversionError in FileHandler.match?

### ISSUE THREAD

**Issue by lemonez** [rails/rails#19187]
I upgraded to Redmine 3.0.0.0 and Rails 4.2.0. I'm using PostgreSQL database adapter, Ruby 1.9.3 and Windows Server 2008.

For some reason I get an error when trying to download files with Cyrillic names that contain a certain character (И).

The line that raises the exception: https://github.com/rails/rails/blob/4-2-stable/actionpack/lib/action_dispatch/middleware/static.rb#L34

Error trace:

    "\x98" to UTF-8 in conversion from Windows-1251 to UTF-8
    c:/ruby193/lib/ruby/gems/1.9.1/gems/actionpack-4.2.0/lib/action_dispatch/middleware/static.rb:34:in `file?'
    c:/ruby193/lib/ruby/gems/1.9.1/gems/actionpack-4.2.0/lib/action_dispatch/middleware/static.rb:34:in `block in match?'
    c:/ruby193/lib/ruby/gems/1.9.1/gems/actionpack-4.2.0/lib/action_dispatch/middleware/static.rb:31:in `each'
    c:/ruby193/lib/ruby/gems/1.9.1/gems/actionpack-4.2.0/lib/action_dispatch/middleware/static.rb:31:in `detect'
    c:/ruby193/lib/ruby/gems/1.9.1/gems/actionpack-4.2.0/lib/action_dispatch/middleware/static.rb:31:in `match?'
    c:/ruby193/lib/ruby/gems/1.9.1/gems/actionpack-4.2.0/lib/action_dispatch/middleware/static.rb:107:in `call'
    c:/ruby193/lib/ruby/gems/1.9.1/gems/rack-1.6.0/lib/rack/sendfile.rb:113:in `call'
    c:/ruby193/lib/ruby/gems/1.9.1/gems/railties-4.2.0/lib/rails/engine.rb:518:in `call'
    c:/ruby193/lib/ruby/gems/1.9.1/gems/railties-4.2.0/lib/rails/application.rb:164:in `call'
    c:/ruby193/lib/ruby/gems/1.9.1/gems/railties-4.2.0/lib/rails/railtie.rb:194:in `public_send'
    c:/ruby193/lib/ruby/gems/1.9.1/gems/railties-4.2.0/lib/rails/railtie.rb:194:in `method_missing'
    c:/ruby193/lib/ruby/gems/1.9.1/gems/thin-1.5.1/lib/thin/connection.rb:81:in `block in pre_process'
    c:/ruby193/lib/ruby/gems/1.9.1/gems/thin-1.5.1/lib/thin/connection.rb:79:in `catch'
    c:/ruby193/lib/ruby/gems/1.9.1/gems/thin-1.5.1/lib/thin/connection.rb:79:in `pre_process'
    c:/ruby193/lib/ruby/gems/1.9.1/gems/thin-1.5.1/lib/thin/connection.rb:54:in `process'
    c:/ruby193/lib/ruby/gems/1.9.1/gems/thin-1.5.1/lib/thin/connection.rb:39:in `receive_data'
    c:/ruby193/lib/ruby/gems/1.9.1/gems/eventmachine-1.0.3-x86-mingw32/lib/eventmachine.rb:187:in `run_machine'
    c:/ruby193/lib/ruby/gems/1.9.1/gems/eventmachine-1.0.3-x86-mingw32/lib/eventmachine.rb:187:in `run'
    c:/ruby193/lib/ruby/gems/1.9.1/gems/thin-1.5.1/lib/thin/backends/base.rb:63:in `start'
    c:/ruby193/lib/ruby/gems/1.9.1/gems/thin-1.5.1/lib/thin/server.rb:159:in `start'
    c:/ruby193/lib/ruby/gems/1.9.1/gems/thin-1.5.1/lib/thin/controllers/controller.rb:86:in `start'
    c:/ruby193/lib/ruby/gems/1.9.1/gems/thin-1.5.1/lib/thin/runner.rb:187:in `run_command'
    c:/ruby193/lib/ruby/gems/1.9.1/gems/thin-1.5.1/lib/thin/runner.rb:152:in `run!'
    c:/ruby193/lib/ruby/gems/1.9.1/gems/thin-1.5.1/bin/thin:6:in `<top (required)>'
    c:/Ruby193/bin/thin:23:in `load'
    c:/Ruby193/bin/thin:23:in `<main>'

**Comment by sivsushruth** [comment]
Hi, 
You can use this gem : https://github.com/nurse/iconv 

```
filename = Iconv.conv('UTF8', 'CP1251', filename)
```

Let me know if this works

**Comment by lemonez** [comment]
@sivsushruth That does not look like what is needed here.
As a workaround I added 

```
rescue Encoding::UndefinedConversionError
  false
```
Very simple, and works for me. Redmine resolves the file name correctly.

Redmine attachment model has two properties: display file name in user defined encoding, and disk file name in ASCII. The problem itself is not that the file name is not converted, but that `match?` is called somewhere before Redmine actually uses the disk file name.

**Comment by pixeltrix** [comment]
@lemonez are you sure that character is correct - looking at the [Window 1251][1] code page shows a grey space indicating an undefined character and [the mapping on the Unicode website][2] confirms that. Is it possible that the filename encoding has been corrupted?

[1]: https://msdn.microsoft.com/en-gb/goglobal/cc305144.aspx
[2]: http://www.unicode.org/Public/MAPPINGS/VENDORS/MICSFT/WINDOWS/CP1251.TXT

**Comment by pixeltrix** [comment]
@lemonez looking at [the encoding info on that character][1] it seems that the filename is in UTF-8 (И is 0xD098 in UTF-8) and not Windows 1251 - where is that encoding being set? Is it the default external encoding?

[1]: http://en.wikipedia.org/wiki/I_(Cyrillic)

**Comment by lemonez** [comment]
@pixeltrix The default external encoding is UTF-8 (I used `raise Encoding.default_external.to_s` to check). Also, database encoding is Unicode.
I'm not sure if this is relevant, but Redmine lets you set up _Attachments and repositories encodings_, and there, I have _UTF-8, CP1251_.

**Comment by alanliu006** [comment]
@lemonez I met exactly the same problem.
Can you please tell where to put the workaround? I have tried to revise static.rb, but not work. Thank you.

      if match = paths.detect { |p|
        path = File.join(@root, p)
        begin
          File.file?(path) && File.readable?(path)
          rescue SystemCallError
            false
		  rescue Encoding::UndefinedConversionError
		    false
        end
      }

**Comment by lemonez** [comment]
@alanliu006 Yes, that's what I have in static.rb

**Comment by alanliu006** [comment]
@lemonez Thanks for the confirmation. Following revise works for me.

      if match = paths.detect { |p|
        path = File.join(@root, p)
        begin
          File.file?(path) && File.readable?(path)
          rescue SystemCallError
            false
          rescue EncodingError
            false
        end
      }

**Comment by marutosi** [comment]
See https://github.com/macournoyer/thin/issues/268#issuecomment-86508080 .

**Comment by marutosi** [comment]
I have send pull request #19533.

**Comment by marutosi** [comment]
Why does not Rails team respond to #19187 and #19533 ?
These are serious Rails 4 regression for Redmine Bitnami Windows users.
https://community.bitnami.com/t/problems-with-3-0-1-installation-see-report-inside/30195/

#19187 is caused on any Windows thin and puma users.
It is **not** caused on webrick users.

Related:
* https://github.com/rack/rack/issues/732#issuecomment-67677272
* https://github.com/phusion/passenger/issues/1328

**Comment by rafaelfranca** [comment]
Merged at 4df216c
